bridge/ffi/src/c_api.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{bindings::*, utils::*};
+use dsnp_graph_config::SchemaId;
+use dsnp_graph_core::{
+	api::api::{GraphAPI, GraphState},
+	dsnp::dsnp_types::DsnpUserId,
+};
+use std::{
+	mem::ManuallyDrop,
+	panic,
+	sync::{Arc, Mutex},
+};
+
+#[no_mangle]
+pub extern "C" fn print_hello_graph() {
+	println!("Hello, Graph!");
+}
+
+// Collection of GraphStates
+static mut GRAPH_STATES: Option<Arc<Mutex<Vec<*mut GraphState>>>> = None;
+
+// Initialize GraphState
+#[no_mangle]
+pub unsafe extern "C" fn initialize_graph_state(
+	environment: *const Environment,
+) -> *mut GraphState {
+	// use panic_unwind
+	let result = panic::catch_unwind(|| {
+		let environment = &*environment;
+		let rust_environment = environment_from_ffi(environment);
+		let graph_state = Box::into_raw(Box::new(GraphState::new(rust_environment)));
+
+		// Initialize GRAPH_STATES vector if GRAPH_STATES is None
+		if GRAPH_STATES.is_none() {
+			GRAPH_STATES = Some(Arc::new(Mutex::new(Vec::new())));
+		}
+		// Add state pointer to GRAPH_STATES vector if GRAPH_STATES is Some
+		GRAPH_STATES.as_ref().unwrap().lock().unwrap().push(graph_state);
+
+		graph_state
+	});
+	match result {
+		Ok(graph_state) => graph_state,
+		Err(_) => std::ptr::null_mut(),
+	}
+}
+
+// Intialize GraphState with capacity
+#[no_mangle]
+pub unsafe extern "C" fn initialize_graph_state_with_capacity(
+	environment: *const Environment,
+	capacity: usize,
+) -> *mut GraphState {
+	let result = panic::catch_unwind(|| {
+		let environment = &*environment;
+		let rust_environment = environment_from_ffi(environment);
+		let graph_state =
+			Box::into_raw(Box::new(GraphState::with_capacity(rust_environment, capacity)));
+
+		// Initialize GRAPH_STATES vector if GRAPH_STATES is None
+		if GRAPH_STATES.is_none() {
+			GRAPH_STATES = Some(Arc::new(Mutex::new(Vec::new())));
+		}
+		// Add state pointer to GRAPH_STATES vector if GRAPH_STATES is Some
+		GRAPH_STATES.as_ref().unwrap().lock().unwrap().push(graph_state);
+
+		graph_state
+	});
+	match result {
+		Ok(graph_state) => graph_state,
+		Err(_) => std::ptr::null_mut(),
+	}
+}
+
+// Get Capacity
+#[no_mangle]
+pub unsafe extern "C" fn get_graph_capacity(graph_state: *mut GraphState) -> usize {
+	if graph_state.is_null() {
+		return 0
+	}
+	let graph_state = &mut *graph_state;
+	graph_state.capacity()
+}
+
+// Get total graph states in GRAPH_STATES
+#[no_mangle]
+pub unsafe extern "C" fn get_graph_states_count() -> usize {
+	if GRAPH_STATES.is_none() {
+		return 0
+	}
+	let graph_states = GRAPH_STATES.as_ref().unwrap().lock().unwrap();
+	graph_states.len()
+}
+
+// Get graph state from GRAPH_STATES at a given index
+#[no_mangle]
+pub unsafe extern "C" fn get_graph_state_at_index(index: usize) -> *mut GraphState {
```

Comment:
This is why I suggested a HashMap instead of a Vec to store GraphState... index is kind of useless, as it may change as a result of an intervening call to `free_graph_state()`.

I would say either switch to a HashMap, or get rid of this function; we probably don't need it anyway, since we never return the index to the client. If we want to be able to iterate over all `GraphState`s, we should just provide access to an iterator.